parsers/jscript.c

```diff
@@ -2520,6 +2524,78 @@ static bool parseLine (tokenInfo *const token, bool is_inside_class)
 				break;
 		}
 	}
+	else if (isType (token, TOKEN_OPEN_PAREN) ||
+			 isType (token, TOKEN_BANG))
+	{
+		/*
+		 * Try to handle "Self-Invoking Anonymous Function Statements" like:
+		 * (function(x) { # case A.
+		 *     return x + 1;
+		 * })(1);
+		 *
+		 * (function(y) { # case B.
+		 *     return y - 1;
+		 * }(1));
+		 *
+		 * !function(z) { # case C.
+		 *     return z * 1;
+		 * }(1);
+		 */
+		bool started_with_paren = isType (token, TOKEN_OPEN_PAREN);
+
+		readToken (token);
+		is_terminated = parseLine (token, is_inside_class);
+
+		/*
+		 * case C is nothing special; t just returns above `is_terminated'.
+		 * In case A, next ')' should be consumed here.
+		 * In case B, the last ')' after an argument list should be consumed here.
+		 * In case A and B, this function returns true.
+		 */
+		if (started_with_paren)
+		{
+			if (is_terminated)
+			{
+				tokenType last_token_type = token->type;
+				/* case A and B. */
+				readToken (token);
+				if (isType (token, TOKEN_CLOSE_PAREN))
+				{
+					/* case A. */
+					is_terminated = true;
+				}
+				else if (last_token_type == TOKEN_CLOSE_CURLY && isType (token, TOKEN_OPEN_PAREN))
+				{
+					/*
+					 * maybe case B.
+					 * Parse the argument list if there is.
+					 */
+					is_terminated = parseStatement (token, is_inside_class);
+					if (is_terminated)
+					{
+						readToken (token);
+						if (!isType (token, TOKEN_CLOSE_PAREN) && !isType (token, TOKEN_EOF))
+						{
+							/*
+							 * In case B, ')' terminates the statement.
+							 * If the token is not ')', the statement is
+							 * not a "Self-Invoking Anonymous Function Statement".
+							 */
+							Assert (NextToken == NULL);
+							NextToken = newToken ();
+							copyToken (NextToken, token, false);
+						}
+					}
+				}
+				else if (!isType (token, TOKEN_EOF))
+				{
+					Assert (NextToken == NULL);
+					NextToken = newToken ();
+					copyToken (NextToken, token, false);
+				}
+			}
+		}
+	}
```

Comment:
I don't really fancy having all this in `parseLine()` instead of `parseStatement()`, but I must admit that it's a lot easier that way, so why not.

Reply:
..I recognize putting all new code to parseLine is not good.
At least I should introduce a new function to handle `(function ()...)`.

I will update my patch.